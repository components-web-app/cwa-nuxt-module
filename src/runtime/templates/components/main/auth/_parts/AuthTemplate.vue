<template>
  <div class="cwa:flex cwa:min-h-full cwa:flex-col cwa:justify-center cwa:py-12 cwa:sm:px-6 cwa:lg:px-8 cwa:bg-dark">
    <div class="cwa:sm:mx-autocwa:sm::w-full cwa:sm::max-w-md">
      <CwaLogo class="cwa:mx-auto cwa:h-12 cwa:w-auto cwa:text-neutral-400" />
    </div>

    <div class="cwa:mt-8 cwa:sm:mx-auto cwa:sm:w-full cwa:sm:max-w-md cwa:text-white">
      <div class="cwa:bg-neutral-800 cwa:py-8 cwa:px-4 cwa:shadow cwa:sm:px-10">
        <CwaUiAlertWarning
          v-if="error"
          class="cwa:mb-4"
        >
          {{ error }}
        </CwaUiAlertWarning>
        <form
          ref="form"
          action="#"
          method="post"
          class="cwa:flex cwa:flex-col cwa:gap-y-6"
          @submit.prevent="$emit('submit')"
        >
          <slot />
          <div v-if="!hideSubmit">
            <button
              type="submit"
              :disabled="submitting"
              :class="{ 'cwa:opacity-50': submitting }"
              class="cwa:cursor-pointer cwa:flex cwa:w-full cwa:justify-center cwa:bg-neutral-600 cwa:py-2 cwa:px-3 cwa:text-sm cwa:font-semibold cwa:text-white cwa:shadow-sm cwa:hover:bg-neutral-500 cwa:focus-visible:outline cwa:focus-visible:outline-2 cwa:focus-visible:outline-offset-2 cwa:focus-visible:outline-neutral-600"
            >
              {{ submitButtonText }}
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { useTemplateRef } from 'vue'

defineEmits(['submit'])
defineProps<{
  submitButtonText: string
  submitting: boolean
  error?: string | null | string[]
  hideSubmit?: boolean
}>()

const form = useTemplateRef('form')

function submitForm() {
  if (!form.value) {
    return
  }
  form.value.submit()
}

defineExpose({
  submitForm,
})
</script>
