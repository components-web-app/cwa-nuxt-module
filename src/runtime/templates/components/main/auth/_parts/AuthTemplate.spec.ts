// @vitest-environment happy-dom
import { describe, expect, test } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import AuthTemplate from './AuthTemplate.vue'

interface LoginMeta {
  submitButtonText: string
  submitting: boolean
  error?: string
  slotContent?: string
}

function createWrapper({
  submitButtonText,
  submitting,
  error = '',
  slotContent = '',
}: LoginMeta) {
  return shallowMount(AuthTemplate, {
    props: {
      submitButtonText,
      submitting,
      error,
    },
    slots: {
      default: slotContent,
    },
    global: {
      renderStubDefaultSlot: true,
    },
  })
}

describe('LoginPage', () => {
  describe('emits', () => {
    test('should emit on form submit', async () => {
      const wrapper = createWrapper({
        submitting: true,
        submitButtonText: 'Submit',
      })
      const form = wrapper.find('form')

      await form.trigger('submit')

      expect(wrapper.emitted('submit')?.length).toEqual(1)
    })
  })

  describe('snapshots', () => {
    test('should match snapshot IF form is submitting', () => {
      const wrapper = createWrapper({
        submitting: true,
        submitButtonText: 'Submit',
        slotContent: '<div>Test form content</div>',
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should match snapshot IF form has errors', () => {
      const wrapper = createWrapper({
        submitting: false,
        submitButtonText: 'Submit',
        error: 'Something went wrong, try submitting form again',
        slotContent: '<div>Test form content</div>',
      })

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
