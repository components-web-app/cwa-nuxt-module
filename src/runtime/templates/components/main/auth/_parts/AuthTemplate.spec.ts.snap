// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > snapshots > should match snapshot IF form has errors 1`] = `
<div
  class="cwa:flex cwa:min-h-full cwa:flex-col cwa:justify-center cwa:py-12 cwa:sm:px-6 cwa:lg:px-8 cwa:bg-dark"
>
  <div
    class="cwa:sm:mx-autocwa:sm::w-full cwa:sm::max-w-md"
  >
    <logo-stub
      class="cwa:mx-auto cwa:h-12 cwa:w-auto cwa:text-neutral-400"
    />
  </div>
  <div
    class="cwa:mt-8 cwa:sm:mx-auto cwa:sm:w-full cwa:sm:max-w-md cwa:text-white"
  >
    <div
      class="cwa:bg-neutral-800 cwa:py-8 cwa:px-4 cwa:shadow cwa:sm:px-10"
    >
      <warning-stub
        class="cwa:mb-4"
      >
        Something went wrong, try submitting form again
      </warning-stub>
      <form
        action="#"
        class="cwa:flex cwa:flex-col cwa:gap-y-6"
        method="post"
      >
        
        <div>
          Test form content
        </div>
        
        <div>
          <button
            class="cwa:cursor-pointer cwa:flex cwa:w-full cwa:justify-center cwa:bg-neutral-600 cwa:py-2 cwa:px-3 cwa:text-sm cwa:font-semibold cwa:text-white cwa:shadow-sm cwa:hover:bg-neutral-500 cwa:focus-visible:outline cwa:focus-visible:outline-2 cwa:focus-visible:outline-offset-2 cwa:focus-visible:outline-neutral-600"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`LoginPage > snapshots > should match snapshot IF form is submitting 1`] = `
<div
  class="cwa:flex cwa:min-h-full cwa:flex-col cwa:justify-center cwa:py-12 cwa:sm:px-6 cwa:lg:px-8 cwa:bg-dark"
>
  <div
    class="cwa:sm:mx-autocwa:sm::w-full cwa:sm::max-w-md"
  >
    <logo-stub
      class="cwa:mx-auto cwa:h-12 cwa:w-auto cwa:text-neutral-400"
    />
  </div>
  <div
    class="cwa:mt-8 cwa:sm:mx-auto cwa:sm:w-full cwa:sm:max-w-md cwa:text-white"
  >
    <div
      class="cwa:bg-neutral-800 cwa:py-8 cwa:px-4 cwa:shadow cwa:sm:px-10"
    >
      <!--v-if-->
      <form
        action="#"
        class="cwa:flex cwa:flex-col cwa:gap-y-6"
        method="post"
      >
        
        <div>
          Test form content
        </div>
        
        <div>
          <button
            class="cwa:opacity-50 cwa:cursor-pointer cwa:flex cwa:w-full cwa:justify-center cwa:bg-neutral-600 cwa:py-2 cwa:px-3 cwa:text-sm cwa:font-semibold cwa:text-white cwa:shadow-sm cwa:hover:bg-neutral-500 cwa:focus-visible:outline cwa:focus-visible:outline-2 cwa:focus-visible:outline-offset-2 cwa:focus-visible:outline-neutral-600"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
